README.md

```diff
@@ -18,6 +18,15 @@ of the pipeline description, and more.
  * **Mailing List**: image-builder@redhat.com
  * **Changelog**: <https://github.com/osbuild/osbuild/releases>
 
+### Principles
+
+1. OSBuild stages are never broken, only deprecated. The same manifest should always produce the same output.
+2. OSBuild stages should be explicit whenever possible instead of e.g. relying on the state of the tree.
```

Comment:
I think this one is very idealistic, which is nice, but there is some change going on that will make this harder in the future.

The biggest change being allowing a large part of the input to be something that isn't from a previous stage. We are now allowing a "bring your own OS Container workflow" in which a large part of the "tree" comes from something external like a container (coreos or bootc based). In these cases, where things can be picked up from the tree I think it should be possible. 

For example, what if I need to run a stage that does something that depends on the version of the kernel (i.e. file path or something to a file and it has the kernel version in it). Previously if we're building everything with OSBuild (including the input OSTree/tree directly from RPMs) then we know the kernel version and it's not a big deal to include that in the manifest. But now, the kernel RPM was baked into the container that was built somewhere else. Now one of two things has to happen:

1. The thing that generates the manifest that is input to OSBuild needs to crack open the container to figure out the kernel version. This can be wasteful IMO as otherwise the container image wouldn't even need to be pulled by the thing that generated the manifest.
2. OSBuild needs to be able to inspect the tree (i.e. to find the kernel version that was used) and drop *some* of its explicit requirements.

Reply:
The way I see it, the exceptions you mention are covered by the "whenever possible" part of this rule.  There are indeed times when it's not possible to write a stage without inspecting the tree.  Sometimes we need to read a configuration file, or modify one without affecting the config keys we're not modifying.  Sometimes we need to find the kernel (like we had with zipl.inst).
But the "whenever possible" part also works the other way: if knowledge of the tree contents is available beforehand, the manifest generator should do everything it can to put that info in the manifest and be explicit.  In the case of containers providing the tree, there are a few things going on now that relate to this:
1. Tools like `bootc` take care of inspecting the tree to do the right thing, so the manifest options aren't involved.  This is, in some ways, the ideal case for stages: thin wrappers around OS utilities.  If a utility _needs_ a certain value to be set, it's not often the case that populating that value requires inspection of the tree and can't be inferred from the stages leading up to it (in my experience anyway).  And other times, when the state can be read from the tree (e.g. a config needs to be read), the tool should be doing that when running.
2. We're enabling using containers from local storage, which means that manifest generators (osbuild-composer and bootc-image-builder) will be able to inspect or run a container to collect information during manifest generation then osbuild can read the same container from the local store, avoiding the need to pull the container twice.  This isn't a perfect solution though.  In the service this wont be possible (the manifest generation and the build run on different machines), but it will be a great improvement for on-prem case.

In the end, if it's not possible (or extremely inefficient) to collect all info during manifest generation, we read the tree.  If the argument for reading the tree however is that it's inconvenient to write the manifest in that way (needing to repeat options for example), then we don't consider those conveniences as valid reasons for violating this rule.

Reply:
> In the end, if it's not possible (or extremely inefficient) to collect all info during manifest generation, we read the tree. 

OK yeah I think this is the key point for me. If I would be able to generate the manifest (using XYZ manifest generation tool) without having to pull and inspect the container image then I really want to not have to do that. If we require that then the manifest generation tool now needs a lot more smarts (container registry pulling, image inspection, knowledge of how to inspect the FS of the container, etc) than if it's just able to point to a container ref in a registry.

Reply:
Also, "whenever possible" might be a little strong of wording. Technically anything is possible :). Maybe we could decrease the strength of the wording OR we could name exceptions for when we lean the other way.

Reply:
I think you are right @dustymabe, that pulling a container in the manifest generation tool is clearly wrong. However, if we can for instance pull just the metadata, that might be ok. Compare to how we treat RPMs: we would never pull the actual RPMs into the manifest generation tool, but we do pull the yum metadata and use that to generate the right manifest.

So if we have the choice between putting some information needed to make an explicit manifest in the container metadata (available at manifest generation time), that would be preferable to having the same data only available in the container itself, so we would have to rummage around at build time.

In case this seems like an arbitrary principle, this was very much informed by the state of affairs when we started creating osbuild. All the tools we used (grub, selinux, lvm, ...) were built around rummaging around on the running machine to figure out hat to do. This made it extremely hard to debug, extremely fragile to changes in those tools, and hard to reuse the components beyond the precise use cases they were originally built for. When we managed to tear these tools apart a bit and decompose them into pieces that did not depend on the tree they were modifying it really helped us a lot. Especially when it came to reusing components for building more than one type of image (it took us months to get RHEL8 to work, but only days for RHEL7, etc).

Reply:
> I think you are right @dustymabe, that pulling a container in the manifest generation tool is clearly wrong. However, if we can for instance pull just the metadata, that might be ok.

I think that's a reasonable and practical middle ground.

Reply:
I've not adjusted the wording on this principle based on the discussion - I hope I read the room correctly :innocent: 
Also, principles are always aspirational! :star2: